mne_connectivity/spectral/time.py

```diff
@@ -133,21 +150,26 @@ def spectral_connectivity_time(
         Amount of time to consider as padding at the beginning and end of each
         epoch in seconds. See Notes for more information.
     mode : str
-        Time-frequency decomposition method. Can be either: 'multitaper', or
-        'cwt_morlet'. See :func:`mne.time_frequency.tfr_array_multitaper` and
-        :func:`mne.time_frequency.tfr_array_morlet` for reference.
+        Time-frequency decomposition method. Can be either: ``'multitaper'``, or
+        ``'cwt_morlet'``. See :func:`mne.time_frequency.tfr_array_multitaper` and
+        :func:`mne.time_frequency.tfr_array_morlet` for reference. Ignored if ``data``
+        is an :class:`~mne.time_frequency.EpochsTFR` object.
     mt_bandwidth : float | None
-        Product between the temporal window length (in seconds) and the full
-        frequency bandwidth (in Hz). This product can be seen as the surface
-        of the window on the time/frequency plane and controls the frequency
-        bandwidth (thus the frequency resolution) and the number of good
-        tapers. See :func:`mne.time_frequency.tfr_array_multitaper`
-        documentation.
+        Product between the temporal window length (in seconds) and the full frequency
+        bandwidth (in Hz). This product can be seen as the surface of the window on the
+        time/frequency plane and controls the frequency bandwidth (thus the frequency
+        resolution) and the number of good tapers. See
+        :func:`mne.time_frequency.tfr_array_multitaper` documentation. If ``data`` is an
+        :class:`~mne.time_frequency.EpochsTFR` object computed with the ``multitaper``
+        method, this should match the value used to compute the TFR.
     n_cycles : float | array_like of float
-        Number of cycles in the wavelet, either a fixed number or one per
-        frequency. The number of cycles ``n_cycles`` and the frequencies of
-        interest ``cwt_freqs`` define the temporal window length. For details,
-        see :func:`mne.time_frequency.tfr_array_morlet` documentation.
+        Number of cycles in the wavelet, either a fixed number or one per frequency. The
+        number of cycles ``n_cycles`` and the frequencies of interest ``freqs`` define
+        the temporal window length. For details, see
+        :func:`mne.time_frequency.tfr_array_multitaper` and
+        :func:`mne.time_frequency.tfr_array_morlet` documentation. If ``data`` is an
+        :class:`~mne.time_frequency.EpochsTFR` object computed with the ``multitaper``
+        method, this should match the value used to compute the TFR.
```

Comment:
statements like this worry me, because they mean that it's easy for people to make mistakes.

For things that are ignored if you pass an `EpochsTFR` object, then (assuming the default param value is `None` or some other null/auto value) we at least have the option of checking whether non-default input to this function *matches* what we extract/infer from the object, and raising warnings/errors if they don't match. But for params that have int/float default values, or for quantities/settings that were used to create the TFR object but aren't stored with it, we can't even do that sanity check.

I don't want to throw a big monkey wrench into this PR, but an API like this makes me think that there should be separate functions for operating on MNE Objects and on arrays, and the shared code path should be a helper func rather than the user-facing function(s). Another possibility is 2 public-facing funcs and no shared helper; one of the public funcs would then call the other after suitable transformation of its inputs.  This could mean internally extracting arrays from objects and always operating on arrays, or it could mean internally promoting arrays to objects and always operating on objects.

Without looking closely I'm not sure how much work that would be / how (in)consistent it is with the rest of the mne_connectivity API.  I also apologize if this was discussed in advance and I missed/forgot about that discussion...  I've been rather underwater for the last several months with other responsibilities.  But I'd like to at least check with you and @larsoner (and maybe @adam2392 if he has time) before this becomes something that needs a deprecation cycle to change.

(side note re: the discussion elsewhere in this PR between @tsbinns and @larsoner about storing estimation params like `n_cycles` as part of the TFR object --- that helps prevent some mistakes, but I still have the feeling like this function is trying to do too much.)

Reply:
> statements like this worry me, because they mean that it's easy for people to make mistakes.

I 100% agree. With storing the params (or the weights directly like discussed in https://github.com/mne-tools/mne-python/issues/12851), then this statement could be removed and prevent possible mistakes.

> but I still have the feeling like this function is trying to do too much.

But yeah, I also understand.

> an API like this makes me think that there should be separate functions for operating on MNE Objects and on arrays

Just to clarify, this would mean something like:
- `spec_conn_epochs/time` handling `Epochs` and `EpochsSpectrum/TFR` objects
- and a `spec_conn_epochs/time_array` function handling array of timeseries data or coefficients

If there were separate functions, would it not be simpler to maintain `spec_conn_epochs/time` as handling timeseries data (either as `Epochs` or arrays), and introduce equivalent functions for handling spectral coefficients (either as `EpochsSpectrum`, `EpochsTFR`, or arrays)? I feel like this would be more compatible with the suggestion below:

> Another possibility is 2 public-facing funcs and no shared helper; one of the public funcs would then call the other after suitable transformation of its inputs. This could mean internally extracting arrays from objects and always operating on arrays, or it could mean internally promoting arrays to objects and always operating on objects.

However for the second sentence, I think again it's less a case of distinguishing data in arrays vs. MNE objects, and more distinguishing timeseries vs. coefficients.

> Without looking closely I'm not sure how much work that would be / how (in)consistent it is with the rest of the mne_connectivity API.

I imagine the most important thing is for whatever is changed in `spec_conn_time`, that should be mirrored in `spec_conn_epochs` and vice versa (so things like https://github.com/mne-tools/mne-connectivity/pull/220 would need to be looked at again). I am perfectly happy to work on this (but I also understand that means more work for reviewers).

Reply:
> I think again it's less a case of distinguishing data in arrays vs. MNE objects, and more distinguishing timeseries vs. coefficients.

yeah you're right I was confusing the issue. Just to make sure we're on the same page: am I right that the "flavors" of spectral connectivity you want to support are:

- input: non-spectral (ordinary time series)
    - across: epochs (`spec_conn_epochs`; objects or arrays)
    - across: times (`spec_conn_time`; objects or arrays)
- input: spectral
    - across: epochs (currently `spec_conn_epochs`; maybe a new func?)
domain)
    - across: time (not supported; no time dimension)
- input: spectrotemporal
    - across: epochs (not supported?)
    - across: times (this PR: `spec_conn_time`; maybe a new func? `spec_conn_tfr_time`?)

Reply:
> input: non-spectral (ordinary time series)
> - across: epochs (spec_conn_epochs; objects or arrays)
> - across: times (spec_conn_time; objects or arrays)

Yes.

> input: spectral
> - across: epochs (currently spec_conn_epochs; maybe a new func?)
> domain)
> - across: time (not supported; no time dimension)

Yes.

> input: spectrotemporal
> - across: epochs (not supported?)
> - across: times (this PR: spec_conn_time; maybe a new func? spec_conn_tfr_time?)

This PR was also adding support for spectrotemporal input for `spec_conn_epochs` to match the `cwt_morlet` mode.